### DISCUSSION TITLE
How to connect backrest to an existing restic repo?

### DISCUSSION THREAD

**Discussion by bhlevca** [garethgeorge/backrest#577] [General]
I like very much this project but I have some trouble with configuring it for my setup.

I already have restic repositories and I would like to be able to be able to browse them and restore from them. The repositories were created with the restic CLI.
I am using the docker option and so far I only can see that I can add/initialize a repo. Is there a way to add an already existing one?  If yes, please let me how can that be done or point  me to the right document. Thanks

**Comment by garethgeorge** [comment]
Hey— the add flow will work if you enter configuration for an existing repo too. They’re the same flows :). 

Backrest will explicitly check if a repo already exists at the location before trying to initialize a new one. If the repo does already exist it’ll be imported and your snapshots will be indexed.

**Reply by bhlevca** [reply]
Thanks for your quick reply.
I tried to connect to the repo but it appears that there is a `ssh` problem for `subprocess`: 

`subprocess ssh: Host key verification failed.
Fatal: create repository at sftp:NAME@ADDRESS://REPO failed: Fatal: unable to open repository at sftp:NAME@ADDRESS://REPO: unable to start the sftp session, error: error receiving version packet from server: server unexpectedly closed connection: unexpected EOF
`

If I am trying from CLI the command `restic -r sftp:NAME@ADDRESS://REP snapshots`  it connects properly and I get the snapshot list
Based on the printed error, I assume there is a problem with the subprocess on the docker's web server.  

Reading your document sI see that you mention that one / works better than // . I tried one / and the error is identical

I think that for debugging purposes these should be a way to test restic directly, such like a dialog box where commands are passed to a terminal where we can enter commands like restic -r sftp:NAME@ADDRESS:/REPO snapshot

I gave up on sftp and I mounted the NAS with NFS.  I was able to "add" the repository but I  I don't see much in the Backrest window. The Tree is empty and I was expecting to see the repository tree, but after issuing several commands I realized that is more like a history of commands rather than the result of running them.  Not very intuitive especially since the documentation is very limited.

After finally adding the repository I saw the "command" button. I ran the simple command "snapshots" and the command output is "command: /bin/restic-0.17.1 snapshots -o sftp.args=-oBatchMode=yes"
I see 2 problems here:

- I was expecting to see something like this in the command output:

```
ID        Time                 Host        Tags        Paths
------------------------------------------------------------
43e400e4  2020-08-04 17:08:47  rpbi48gb                /
ff8faa0e  2020-08-24 16:20:22  rpbi48gb                /
80d86e35  2020-09-13 21:34:14  rpbi48gb                /
8de14c5c  2020-10-06 23:59:06  rpbi48gb                /
...
------------------------------------------------------------

```

- why are the following arguments added "-o sftp.args=-oBatchMode=yes"


Also, it appears that there is no way to do a backup unless you have a "Plan". I prefer doing manual backups.  

It seems that you have a very nice project, but due to my limited time and lack of extensive documentation I am unable to make something useful out of it,  I was hoping for something more user-friendly where I connect to the repo and can browse the files, extract files, start manual backups.  Maybe it is possible with this software, but as I said with my limited time and lack of detailed documentation I can't.  Thanks.

**Reply by garethgeorge** [reply]
Sorry you’re having trouble, the docs on the wiki cover much of this first time setup, https://garethgeorge.github.io/backrest . 

It sounds like you are looking for the “index snapshots” button which reads the repo and imports your snapshots if you’re not using a plan to automate backups. 

You can certainly just run manual backups using the “run command” button and then index them with “index snapshots” to see stats and browse files.

**Reply by unari-san** [reply]
> subprocess ssh: Host key verification failed

i had the same issue. could use sftp in the command line without a problem but not in backrest. i tried `-o sftp.args=-i /path/to/key -o StrictHostKeyChecking=no` in backrest but got an error (probably because i used one flag for two args). it did however verify the host key properly. i removed the StrictHostKeyChecking option afterwards and i was able to add the (existing) repo.

> [unknown] command "/root/.local/share/backrest/restic cat config -o sftp.args=-i /path/to/key -o StrictH..." failed: command output is not valid JSON: invalid character 's' looking for beginning of value
> Output:
> subprocess ssh: Warning: Permanently added 'host' (ED25519) to the list of known hosts.
> {
>   "version": 1,
>   "id": "",
>   "chunker_polynomial": ""
> }

**Reply by unari-san** [reply]
probably related to https://github.com/garethgeorge/backrest/discussions/341#discussioncomment-11650794

i'm not using docker though

**Comment by bhlevca** [comment]
I read the wiki but since it does not provide images of how the page should look like, see how mine look in the attachments
The commands you mention are not giving any output to see any snapshots or files.  I am not sure if my setup is correct, but definitely I am not getting any useful output

Also, ssh does not seem to work and even the snapshot command has flags that are unnecessary (seem to be there by default)
Could it be that the docker image faulty, should I try the linux deployment instead? 
In any case the Wiki needs more information and images for a first time user.

![Baxckrest1](https://github.com/user-attachments/assets/8b065d5f-bed6-4eb2-817b-7cba5dda39de)
![Baxckrest2](https://github.com/user-attachments/assets/336ba065-cdc7-4d41-84fa-ee7a2bf92a7a)

**Reply by bhlevca** [reply]
I am expecting to see something like seen in the attached.  It took 2 min to get those results

![ResticBrowser](https://github.com/user-attachments/assets/b741db07-3a3b-4d6e-8e5a-ff5a486f723e)

**Reply by garethgeorge** [reply]
I think what’s probably happening for you is that backrest created a new repo and isn’t seeing anything in it — probably because of a slight typo in the URI or similar. 

I think you have a good point that there’s a lot that can be confusing here, just tried this on a new laptop and it can be tricky to notice as backrest happily just initializes the repo for you if it doesn’t exist. 

I’ll file a bug to take a look at showing a “confirmation” or similar before ever initializes a new repo.

**Comment by bhlevca** [comment]
Also please have a look at SFTP settings. It doesn't work, the subprocess
is looking for some keys it hasn't access to.

I didn't make a typo in the repo path.  I used NFS, perhaps that's a
problem? 
I used the same setting with the restic browser as you can see in the images posted above.

However, I suspect that adding a repo with using the init command
does not work. A separate command to access an existing repo will add
clarity and avoid issues like this one.

**Reply by garethgeorge** [reply]
Hmm, Backrest should not be running init on your repo if it already exists, it first runs `restic cat config` per the docs https://restic.readthedocs.io/en/latest/075_scripting.html#check-if-a-repository-is-already-initialized here to check if a repo is already initialized at the configured location.

I agree though that either way, I think you're right it's confusing that backrest will implicitly initialize without making it clear it's doing this.

I'm curious -- using the `Run Command` feature with `snapshots` do you see any output at all? From your initial bug it sounds like perhaps not? You also asked a question there about `-o sftp.args=-oBatchMode=yes` -- this is a default arg backrest adds which applies to sftp users. It disables any interactive prompting from sftp i.e. waiting for a password over stdin. I'd also add that SFTP setup is _much_ easier when not using docker (i.e. a bare metal install). I am considering disabling it or adding warnings for docker users due to the number of relatively difficult footguns around correctly mounting ssh keys into your container (with the right ownership and permissions) and getting them to be detected by the ssh subprocess forked by restic.

**Comment by bhlevca** [comment]
Thanks for your reply. I don't have the docker running anymore, but I remember that when I ran the `snapshot`  command it did not list anything.  I should probably run the linux version but the whole point of choosing your solution over the Restic Browser or others, is the docker implementation.  Perhaps if you make the docker installation document more detailed about how to configure the docker solution, in particular the SFTP keys, I will try again.

**Comment by quanltt** [comment]
I would like to share my personal experience of using an existing repository and snapshots created by the cli with Backrest. 

In principle, backrest uses tags to organize snapshots (to know which snapshots to be used with what instance and what plan). Two tags are needed: created-by:<instance-id>, plan:<plan-name>.  You only need to create appropriate tags for each snapshot, after which you can reuse them. 

My use case is that I previously backed up my folders using the CLI, and now I want to migrate everything to Backrest. So the source folders are the same, and I can reuse all old snapshots for my new backup in backrest. So here are my steps

1. When declaring the repository, use exact repository uri and pass of an existing one
2. After the repository is recognized and registered with backrest (having a blue tick), click on the repository, and then on the right, click on Check Now. After these steps, I get all the old snapshots into _unassociate_ tree in tree view. These snapshots can be browsed, etc...

These snapshots seem not to be ready to use for the backup using backrest. To use old snapshots with new backup using backrest, I did the following steps

1. Create a backup plan that uses the new repository, with all my backup folders. 
2. Click on the repository in backrest, and click Run Command on the righ, enter this: tag --add created-by:<instance-id> (please do note include <>, but do include : and no blank alow; for example, tag --add created-by:abcd, with abcd is instance id). You can find instance-id in the settings or top right of the screen. Click Excecute and wait. This step adds new tag created-by for all snapshots
4. After the command is finished, click on Run Command again and enter tag --add plan:<plan-name>, for example, tag --add plan:mybackup, then click Execute and wait again. This step adds new tag plan for all snapshots
5. After these steps, the previous snapshots in the repository will be associated with this instance and the plan and will be used for new snapshots by this plan. 

Please note that my use case is simple. I use one backup (with several source folders) to one repository. When I switch to backrest I only need one plan and associate all previous snapshots to that plan. 

Backrest allows more complex settings (many instances, each with many plans).  But in this case, to correctly associate old snapshots with new instance and plan, you have to manually add the two tags to each snapshot or use filters with the tag command (please read restic doc for how to use filter to tag selected snapshots).

**Reply by bhlevca** [reply]
Thanks for the very detailed process of adapting old snapshots to be viewed in backrest. However, this process requires a lot of steps and time, which I don't think that I am going to do since I already have a guy tool "restic browser" that sees the CLI created backups. 
Second, I wanted to use the backrest docker, which has a major problem with SFTP, which disqualified it for me to use.
Third, I expect when I connect to the repo to see all the snapshots without any interventions, it should be fully compatible with the CLI.

I like the fact that I can use a web interface but I am not going to use it until full compatibility with CLI, and SFTP are fixed.

Thanks again

**Reply by garethgeorge** [reply]
Thanks @quanltt for the detailed writeup, that's a great list of steps for importing existing snapshots into backrest's tag system. It's honestly something I'd love to add some GUI support for eventually, but in the meantime I'm also very happy to hear / see that it's relatively straight forward with the Run Command feature :).

Re: backrest support, I do again want to clarify that backrest is fully compatible with the restic CLI (and has test coverage to this effect) and existing repositories. Something is incorrectly configured when you attempt to import your snapshots, and I strongly suspect you've accidentally created a new empty repository somewhere which is why you see no errors and also don't see your data. In https://github.com/garethgeorge/backrest/pull/582 I've added a new button to make it easier to detect this sort of situation. 

SFTP is supported in backrest docker, but requires some amount of work as it would with any docker container as you must _correctly_ mount your key files into the container with appropriate permissions and provide them to backrest.  https://github.com/garethgeorge/backrest/discussions/278 includes some detailed discussion of what's needed for setup.  It's not clear to me that you've taken debugging steps to correct your SSH setup in your docker environment, if you expect things to "just work" a baremetal install is easier here. 

Agreed it would be great to make these flows a bit easier, there are a number of inherent foot guns when setting up SSH in a docker container, but these do work today.

**Comment by bhlevca** [comment]
I don't dispute that "backrest" may eventually be made to work fine. The existent documentation is incomplete for SFTP on docker and it did not cross my mind to dig through discussions.   I am computer savvy and I used to be myself a computer programmer I am sure that eventually I could make it work. However, since I already have a perfectly working solution I am not willing to invest more time. I will perhaps revisit backrest  when it matures.
